cdap-tms/src/main/java/io/cdap/cdap/messaging/store/leveldb/LevelDBTableFactory.java

```diff
@@ -124,6 +139,148 @@ public void close() {
+                                                                                            Bytes.EMPTY_BYTE_ARRAY));
+        }
+      }
+    } catch (IOException ex) {
+      LOG.debug("Unable to perform data cleanup in TMS LevelDB tables", ex);
+    }
+  }
+
+  /**
+   * Delete messages of a {@link TopicId} that has exceeded the TTL or if it belongs to an older generation
+   *
+   * @param now current timestamp
+   * @throws IOException error occurred while trying to delete a row in LevelDB
+   */
+  private void pruneMessages(long now, DB levelDB, TopicMetadata topicMetadata,
+                             Function<Map.Entry<byte[], byte[]>, TableEntry> tableEntryFunc) throws IOException {
+    TopicId topicId = topicMetadata.getTopicId();
+    int topicGeneration = topicMetadata.getGeneration();
+
+    WriteBatch writeBatch = levelDB.createWriteBatch();
+    long ttlMillis = TimeUnit.SECONDS.toMillis(topicMetadata.getTTL());
+    byte[] startRow = MessagingUtils.toDataKeyPrefix(topicId,
+                                                     Integer.parseInt(MessagingUtils.Constants.DEFAULT_GENERATION));
+    byte[] stopRow = Bytes.stopKeyForPrefix(startRow);
+    boolean done = false;
+    int batchSize = 0;
+
+    while (!done) {
+      done = true;
+      try (CloseableIterator<Map.Entry<byte[], byte[]>> rowIterator = new DBScanIterator(levelDB, startRow, stopRow)) {
+        while (rowIterator.hasNext()) {
+          done = false;
+          Map.Entry<byte[], byte[]> entry = rowIterator.next();
+          TableEntry tableEntry = tableEntryFunc.apply(entry);
+
+          int dataGeneration = tableEntry.getGeneration();
+          if (!topicId.equals(tableEntry.getTopicId())) {
+            LOG.warn("Ignore table entry with topic '{}' that doesn't match with the topic '{}'",
+                     tableEntry.getTopicId(), topicId);
+            continue;
+          }
+          if (topicGeneration != dataGeneration) {
+            LOG.warn("Ignore table entry with topic generation '{}' that doesn't match with the topic generation '{}'",
+                     tableEntry.getGeneration(), dataGeneration);
+            continue;
+          }
+
+          if (MessagingUtils.isOlderGeneration(dataGeneration, topicGeneration)) {
+            writeBatch.delete(entry.getKey());
+          } else if ((dataGeneration == Math.abs(topicGeneration)) && ((now - tableEntry.getTimestamp()) > ttlMillis)) {
+            writeBatch.delete(entry.getKey());
+          } else {
+            // terminate scanning table once an entry with write time after TTL is found to avoid scanning whole table,
+            // since the entries are sorted by time.
+            done = true;
+            break;
+          }
+          batchSize++;
+          if (batchSize >= cleanupBatchSize) {
+            break;
+          }
+        }
+      }
+
+      try {
+        levelDB.write(writeBatch, new WriteOptions().sync(true));
```

Comment:
do we need to reinitialize write batch for every batch?

Reply:
You are right. Good catch.